### ISSUE TITLE
webui: empty values lead to faulty 'default.nix'

### ISSUE THREAD

**Issue by tupakkatapa** [ponkila/HomestakerOS#9]
Passing empty or null values through JSON to `default.nix` can result in a faulty configuration:

<details>

<summary> default.nix </summary>

```nix
{
  pkgs,
  config,
  inputs,
  lib,
  ...
}: {
  homestakeros = {
    addons = {
      mev-boost = {
        enable = false;
        endpoint = "http://127.0.0.1:18550";
      };
    };
    consensus = {
      lighthouse = {
        dataDir = "/var/mnt/lighthouse";
        enable = false;
        endpoint = "http://127.0.0.1:5052";
        execEndpoint = "";
        jwtSecretFile = "";
        slasher = {
          enable = false;
          historyLength = 4096;
          maxDatabaseSize = 256;
        };
      };
    };
    execution = {
      erigon = {
        dataDir = "/var/mnt/erigon";
        enable = true;
        endpoint = "http://127.0.0.1:8551";
        jwtSecretFile = "";
      };
    };
    localization = {
      hostname = "foo";
      timezone = "";
    };
    mounts = {};
    ssh = {
      authorizedKeys = [];
      privateKeyFile = "";
    };
    vpn = {
      wireguard = {
        configFile = "";
        enable = false;
        interfaceName = "wg0";
      };
    };
  };
}
```

</details>

Options with empty or null values should be removed from the JSON since they can produce the following types of errors, even if building is possible without setting the `ssh.privateKeyFile` option:

```shell
[I] kari@torque ~/W/HomestakerOS (main) [SIGINT]> tail -f pipe | sh
tail: inotify resources exhausted
tail: inotify cannot be used, reverting to polling
warning: Git tree '/home/kari/Workspace/HomestakerOS' is dirty
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'kexec-tree'
         whose name attribute is located at /nix/store/s4jqyj35hii03rs7j5n6vn7gpgp6ja81-source/pkgs/stdenv/generic/make-derivation.nix:300:7

       … while evaluating attribute 'buildCommand' of derivation 'kexec-tree'

         at /nix/store/s4jqyj35hii03rs7j5n6vn7gpgp6ja81-source/pkgs/build-support/trivial-builders/default.nix:87:14:

           86|       enableParallelBuilding = true;
           87|       inherit buildCommand name;
             |              ^
           88|       passAsFile = [ "buildCommand" ]

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: A definition for option `homestakeros.ssh.privateKeyFile' is not of type `null or path'. Definition values:
       - In `/nix/store/8kfdf8sl274jqd3x7n223axv8wg7vpxy-source/nixosConfigurations/foo': ""
```

**Comment by tupakkatapa** [comment]
Actually, I think that null values are still OK to evaluate and could even have a purpose in the future, but they need to actually be 'null'.. not an empty string.

**Comment by jormeli** [comment]
Should all empty values be left out? Or just ones with the `nullOr` type?

**Comment by tupakkatapa** [comment]
> So `nullOr`-types should be set to `null` if they don't have a value? Does the same apply to other types?

Yes, `nullOr` types can be set to actual `null`. If the type has not `nullOr` in it and value is empty, it should be removed from the JSON altogether so it does not evaluate in the `default.nix`.

**Comment by tupakkatapa** [comment]
In Nix, an empty string does not count as `null`. There is also a possibility that some conversion process between the front and back end will evaluate the `null` values as empty strings, but that is something I will handle if it happens.

**Comment by tupakkatapa** [comment]
Closed by #12.
